python_modules/dagster/dagster/_core/definitions/freshness_checks/time_partition.py

```diff
@@ -84,13 +118,169 @@ def build_time_partition_freshness_checks(
         AssetChecksDefinition: An `AssetChecksDefinition` object, which can execute a freshness
             check for each provided asset.
     """
-    return build_freshness_checks_for_assets(
-        assets=assets,
+    check.str_param(timezone, "timezone")
+    check.opt_str_param(deadline_cron, "deadline_cron")
+    check.invariant(
+        deadline_cron is None or is_valid_cron_string(deadline_cron), "Invalid cron string."
+    )
+    check.inst_param(severity, "severity", AssetCheckSeverity)
+    check.sequence_param(assets, "assets")
+    ensure_no_duplicate_assets(assets)
+    return _build_freshness_multi_check(
+        asset_keys=assets_to_keys(assets),
         deadline_cron=deadline_cron,
         timezone=timezone,
         severity=severity,
-        asset_property_enforcement_lambda=lambda assets_def: check.invariant(
-            isinstance(assets_def.partitions_def, TimeWindowPartitionsDefinition),
-            "Asset is not time-window partitioned.",
-        ),
     )
+
+
+def _build_freshness_multi_check(
+    asset_keys: Sequence[AssetKey],
+    deadline_cron: str,
+    timezone: str,
+    severity: AssetCheckSeverity,
+) -> AssetChecksDefinition:
+    params_metadata: dict[str, Any] = {
+        TIMEZONE_PARAM_KEY: timezone,
+        DEADLINE_CRON_PARAM_KEY: deadline_cron,
+    }
+
+    @multi_asset_check(
+        specs=[
+            AssetCheckSpec(
+                "freshness_check",
+                asset=asset_key,
+                metadata={FRESHNESS_PARAMS_METADATA_KEY: params_metadata},
+                description="Evaluates freshness for targeted asset.",
+            )
+            for asset_key in asset_keys
+        ],
+        can_subset=True,
+        name=f"freshness_check_{unique_id_from_asset_keys(asset_keys)}",
+    )
+    def the_check(context: AssetCheckExecutionContext) -> Iterable[AssetCheckResult]:
+        for check_key in context.selected_asset_check_keys:
+            asset_key = check_key.asset_key
+            current_timestamp = pendulum.now("UTC").timestamp()
+
+            partitions_def = cast(
+                TimeWindowPartitionsDefinition,
+                context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
+            )
+            check.invariant(
```

Comment:
Nit: might be able to make this shorter with the following?
```python
partitions_def = check.inst(
    context.job_def.asset_layer.asset_graph.get(asset_key).partitions_def,
    TimeWindowPartitionsDefinition,
)
```

Reply:
good callout.